### PR TITLE
Fix #134: Ensure Discord is destroyed before IoThread

### LINKED ISSUE #134: IoThreadHolder edge case
Okay, so my situation might be a little bit specific, but perhaps can be solved without requiring me to use `DISCORD_DISABLE_IO_THREAD`. (This is actually the second project I've experienced this issue with, but only now actually figured out exactly what is happening.)

I build the discord-rpc library statically and link it into a dll, which I then dynamically load into a process.

This works great when cleanly shutting down with `Discord_Shutdown()`, but not on process termination. In this case, `~IoThreadHolder()` gets called (because of the static instance of `IoThreadHolder`), which runs `Stop()` which causes a deadlock because it's waiting on a [non-existing thread that has already been terminated](https://msdn.microsoft.com/en-us/library/windows/desktop/ms682658(v=vs.85).aspx):

> All of the threads in the process, except the calling thread, terminate their execution without receiving a DLL_THREAD_DETACH notification.

The MSDN page also states:

> If one of the terminated threads in the process holds a lock and the DLL detach code in one of the loaded DLLs attempts to acquire the same lock, then calling ExitProcess results in a deadlock.

The same thing also happens when trying to call `Discord_Shutdown()` in `DllMain` when `lpvReserved` is non-null, but that can be avoided by simply not calling that, as process termination here is essentially not clean.

I know this might be sort of an edge-case for me, but perhaps this can be fixed somehow?

### PR DESCRIPTION
**Pull request by IntelOrca**
Fix #134: IoThreadHolder edge case

If the static fields are destroyed in the wrong order then a deadlock can occur when calling `Discord_Shutdown`. This aligns the creation and deletion of IoThread with Discord. Ideally IoThread should be a member of connection but that would require moving the definition of it to one of the headers.

This fixes the shutdown issue we were getting in OpenRCT2.
https://github.com/OpenRCT2/OpenRCT2/pull/7431

### CODE REVIEW

**Comment by janisozaur** [comment]
To provide a bit more information: the deadlock mentioned here is one of the ways the issue can exhibit itself and it will very much depend on used toolchain and system.

The real issue we're dealing with here is the static initialisation order fiasco, whereby user of discord-rpc library cannot guarantee `Discord_Shutdown` gets called only before the `IoThreadHolder::~IoThreadHolder`.

File: src/discord_rpc.cpp [hunk 1]
```diff
@@ -346,7 +353,12 @@ extern "C" DISCORD_EXPORT void Discord_Shutdown(void)
     Connection->onConnect = nullptr;
     Connection->onDisconnect = nullptr;
     Handlers = {};
-    IoThread.Stop();
+    if (IoThread != nullptr) {
+        IoThread->Stop();
+        delete IoThread;
+        IoThread = nullptr;
```
Comment (janisozaur) [inline src/discord_rpc.cpp]: It might be worth pointing out this is **not** thread-safe, but the library itself is not meant to be used in such way.

**Comment by MinnDevelopment** [comment]
I believe I also identified a race condition on `Discord_Shutdown` causing `std::thread::join` to be called twice resulting in:

```
libc++abi.dylib: terminating with uncaught exception of type std::__1::system_error: thread::join failed: Invalid argument
```

This could be resolved by only allowing the thread to be stopped once.

**Comment by msciotti** [comment]
Thank you very much for this fix; incredibly helpful. Tagging @snowyote to review!

**Comment by msciotti** [comment]
Sorry for the delay—happy to merge this to unblock folks.

**Pull request merged**